### DISCUSSION TITLE
Parsing paragraphs

### DISCUSSION THREAD

**Discussion by mitko** [mitko/readable_climate_reports#8] [General]
I was able get `pdfjs` to parse all lines and group them into paragraphs - see the attached file. I pushed the code to `mitko/ts_parser` branch.
Some issues:
* this seems to work OK for text, but `pdfjs` is rendering library, not a semantic library, so it chunks text in random chunk sizes.
* there are occasional extra spaces. For example CO2 often uses a sub-script for the 2, which sometimes results in extra " " being added in there. I think this should be cleaned up at a later stage, where we can replace `CO 2` by `CO2` and other such errors.
* The report itself has some weird inconsistencies. For example there are two `FOOTNOTE 9` in the executive summary - the part that's supposed to be the most polished.

@petermr - you mentioned that you were able to extract images? `pdfjs` doesn't seem to be great for images, no single `getImages` API, but many different rendering APIs for types of images. Can you share how you were able to get images? cc @rowanc1 @yipperbore 

Paragraphs from executive summary:
[readableParagraphs.txt](https://github.com/mitko/readable_climate_reports/files/8486191/readableParagraphs.txt)

<img width="816" alt="image" src="https://user-images.githubusercontent.com/125253/163313147-fdaea51d-c3b4-4d26-b12d-e7af9b32f01e.png">

**Comment by petermr** [comment]
Will commit a subset shortly (am reluctant to commit 3000 pages)

**Reply by mitko** [reply]
@petermr - don't commit the raw data or the report itself please, you can put it behind `.gitignore` and have your code generate it locally when run

**Comment by petermr** [comment]
I haven't committed the PDF, only a subset of extracted data. But I can remove this if you'd like.

**Reply by petermr** [reply]
I agree it's a grey area

**Reply by petermr** [reply]
Have removed extracted images from (latest version) of repo.

**Reply by mitko** [reply]
Sounds good! no worries :)

**Reply by petermr** [reply]
I think we need a small number of samples of output - helps to explain and also is permitted as fair use. But we don't want to upset the authors

**Comment by petermr** [comment]
I have written code to extract lines and determine the start of phrases within these.

**Reply by mitko** [reply]
Oh interesting!!

**Comment by yipperbore** [comment]
Try using the bridge report first, it will have a blend of media types and it is the next step from the policy version. If we can get it to work then the 3K pages should be easier.

**Comment by yipperbore** [comment]
No worries’ last night I was researching calls and extensions to help MAI a parsed feed for retrieval. First, I wanted to see if there is a uniform taxonomy, we can apply that would be a little more universal to collect other data types in synthesis. Just concerned if we start with a taxonomy that is custom then we will need to bridge it. Also concerned if it is pure auto driven then retrieval becomes low fidelity. This has been an academic issue forever, trying to bridge lay terms to scientific terms to make the content retrievable for the masses.

In effect you are almost talking two separate languages the IPCC full report is dense with scientific terminology (and a certain etiquette) that most people calling for a better understanding wouldn’t understand how to find information. Here is an example: LCA is a concept, essentially the output of which is a carbon lifecycle for a product. Using a retrieval system if you wanted to know how much carbon products admit without knowing that an LCA even exists?

For services like Google no big deal fuzzy matches and an overload of content guarantees some random variation of search will produce a meaningful (if not factual) document to match your search (this is the beginning of echo chamber formations). In scientific research even for lay people the science must be represented in an approachable manner and the search has to serve as the bridge.

My .02 anyway.

**Reply by mitko** [reply]
> In effect you are almost talking two separate languages the IPCC full report is dense with scientific terminology (and a certain etiquette) that most people calling for a better understanding wouldn’t understand how to find information. Here is an example: LCA is a concept, essentially the output of which is a carbon lifecycle for a product. Using a retrieval system if you wanted to know how much carbon products admit without knowing that an LCA even exists?

Yeah, this is a big problem! I think we can look for some semantic indexing once we get there. I think even starting with executive summary and technical report and reformatting them to be readable will be a big gain

**Reply by petermr** [reply]
There are several steps here. The first and probably easiest is to identify the acronym. , e.g. as https://en.wikipedia.org/wiki/Life-cycle_assessment

**Comment by petermr** [comment]
# Current petermr strategy/activity

I'll outline my immediate activities, which I'm happy to in/decrease according to how things go. I have quite a lot of existing code and strategies, some of which have been developed over several years. They are really only needed because people insist on publishing PDF as he "final" product rather than creating something semantic. Years of trying to convince the world of the value of semantic authoring have made little progress and have active opposition from vested interests. So roughly:

## PDF to syntactic text
By "syntactic" I mean that the structure and components of a document are identified but without any meaning (semantics). A document might consist of (*=repeatable)
```
document
  metadata
  chapter*
    section*
       subsection* (indefinitely nested)
         paragraph*
           sentence*
              phrases/word*
                character*
     figure*
       caption
       bitmap or vectors
     table*
        caption
        row*
           cell*
     reference*
     footnote*
     other_float*
```
"float" is a chunk which does not have to be in logical order.
I will try to recover this for the "finalDraft" .  Don't worry if a syntactic version is produced by the authors - I need to edit the software anyway.

The "syntactic text" is roughly the raw Word/LaTeX document from the publisher and what we can expect in a little while

All components should :
* have explicit syntax (encoding, line-ends, language, markup schemas)
* have unique permanent identifiers
* ideally be normalized against platforms (i.e. behave the same in different OS's). This isn't easy.
* have unicode code-points, normalized by (say) NFC
* use a common encoding ("UTF-8")
* use W3C conventions where possible

Strongly recommend XML and/or HTML for this. It's the publishing standard and flexible enough to hold all the concepts we need

## Syntactic-to-semantic

This is a fluid distinction but generally represents adding metadata that machines can process automatically. Often this can be done with markup languages which define structure, syntax and interpretation. This will probably go beyond what the publishers provide. Examples include:

### JATS-XML for scientific document structure 

JATS-XML is a NISO standard and has ca 250 markup elements which can almost certainly cope with the reports. It's developed by NIH and used by all publishers. It can support other markup languages (see below)

### HTML

HTML is very flexible and can support indefinite nesting `<div>` and very importantly `<span>` , `<sub>` , `<table>` and much else. A mixture of JATS and HTML can support the whole document. There are many tools for rendering HTML to mobiles, epub, accessibility readers, etc.

### Scientific / Technical markup languages

The most important is MathML. We can also have CML for chemistry.

### Linking / Transforming

Linking will be required for several purposes:

* composing a viewable version
* adding inter-chunk links
* linking to outside resources

HTML's `<a href...>` is generally sufficiently powerful (JATS uses W3C's `xlink` which may be preferable.

This section requires agreed heuristics and in some cases manual editing (probably for Math and labelling some sections).

### Ontology (Semantic Web)

This will be where the real power of a readable_document comes. The terms and phrases in the document can be linked to external authorities. For example "CDR" can be linked to `Carbon_dioxide_removal` at https://en.wikipedia.org/wiki/Carbon_dioxide_removal and its Wikidata page at https://www.wikidata.org/wiki/Q56298528 . This also gives us a globally agreed ID for the term and concept.

Where possible we can represent relationships as triples in RDF. This means they will be part of huge global graphs and we can use graph searching tools to extract complex relationships. 

Wikipedia and

**Comment by petermr** [comment]
# petermr's planned work on parsing paragraphs

I have done preliminary tests on the 3000 page finalDraft and it looks quite tractable:

## positives

* no obvious encoding problems
* unicode characters widely used 
* common public fonts (TimesNewRoman, Calibri, etc.)
* almost all figures are RGB-able bitmaps
* expect to be able to trim headers and footers

## unknowns

* not completely clear how/whether spaces are encoded (there are certainly explicit ones)
* How are tables identifiable? (Captions? Gridlines?)
*  Can figure captions be extracted?

## negatives

* some  figures have captions as part of the image. This will require OCR (but we have been developing this recently)

# toolkit

Am using pre-existing tools and some Python to explore the document.

## PDF to SVG
This is a good way of extracting the PDF parsing **with minimal loss**. (We ignore any PDF dictionaries as these may be tool-specific - later they may be useful).

 ### PDFBox-ami3

(This was stable about 3 years ago).

 The PDF is parsed with `PDFBox` running in petermr's `ami3` (https://github.com/petermr/ami3). This:
* reads the PDF pages using PDFBox
* with a subclassed PageViewer, translates the PDFStream into graphical objects (including characters).
* outputs any bitmaps as PNG, labelled by page and x-y coordinates
* outputs graphics primitives as SVG (this mostly occurs in gridded tables)
* outputs characters with text attributes coordinates and style

A typical snippet of output :
```
   <text y="528.124" x="71.998,77.038,79.798,82.558,85.919,90.958,95.998,99.839,102.359,107.398,112.438,115.198,120.118,124.558,127.078,130.438" 
svgx:width="0.5,0.28,0.28,0.33,0.5,0.5,0.39,0.25,0.5,0.5,0.28,0.5,0.44,0.25,0.33,0.72" 
style="fill:rgb(0,0,0);font-family:TimesNewRomanPSMT;font-size:9.96px;stroke:none;">nitrous oxide (N</text>
   <text y="529.08" x="137.64" svgx:width="0.5" style="fill:rgb(0,0,0);font-family:TimesNewRomanPSMT;font-size:6.48px;stroke:none;">2</text>
```
NOTE:
* all coords in screen coords
* all characters with the same y-coord are concatenated, usually in reading order
* the x-coords of each character are given. This will help decide on where the spaces are (a major concern)
* the widths of each character (as pixel-width/font-size ratio). These are also critical for spacing.
* the styling attributes may be critical for guessing the semantics
  - fill (the bulk colour of the text - here black (0,0,0)
  - font-family (often carries semantics)
  - font-size (in screen coords/pixels)
  - stroke (the edge of the character, usually none)
 in addition
* rotated text (not shown here) include transformation matrices

Note there are two text spans here. We have to calculate the length of each (using widths) to determine their display order. When this is done (by `ami3`) we see they are in order. The second has a *lower* baseline (529.1 instead of 528.1 pixels and is smaller (6.48 rather than 9.96). This makes it appear as a subscript - the geometry is the *only* clue. The result in HTML could be:
```
<span font-family="TimesNewRomanPSMT">nitrous oxide (N</span><sub>2</sub>)
```
rendered as flowable

<html>
<span font-family="TimesNewRomanPSMT"
>nitrous oxide (N</span><sub>2</sub><span  font-family="TimesNewRomanPSMT">O); fluorinated gases (HFCs; PFCs, SF</span>
</html>
 
This SVG need only be created once and is easily transformed to HTML (which may be easier).

# likely workplan over next "few days"

## primitives
* Convert all pages to SVG (already proofed - takes ca 0.3 sec / page )
* work out spacing - adjust algorithm
* check extraction of images
* record the process and upload a very few seminal examples of processed pages
* analyse the character attributes throughout the document (size, colour, font-family)

## paragraphs
* create a heuristic paragraph-creator
* create a heuristic labeller where possible
* analyze sub/superscripts
* interpret styles as semantics where possible.

## sub/sections and chapters

* explore what is possible
*
